src/shared/rereadpt.c

```diff
@@ -0,0 +1,355 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <linux/fs.h>
+#include <sys/file.h>
+#include <sys/ioctl.h>
+
+#include "sd-device.h"
+
+#include "alloc-util.h"
+#include "blkid-util.h"
+#include "blockdev-util.h"
+#include "device-util.h"
+#include "device-private.h"
+#include "errno-util.h"
+#include "fd-util.h"
+#include "log.h"
+#include "parse-util.h"
+#include "rereadpt.h"
+#include "set.h"
+#include "string-util.h"
+
+static int fallback_ioctl(sd_device *d, int fd) {
+        assert(d);
+        assert(fd >= 0);
+
+        if (ioctl(fd, BLKRRPART, 0) < 0)
+                return log_device_debug_errno(d, errno, "Failed to reread partition table via BLKRRPART: %m");
+
+        log_device_debug(d, "Successfully reread partition table via BLKRRPART.");
+        return 0;
+}
+
+#if HAVE_BLKID
+static int process_partition(
+                sd_device *d,
+                int fd,
+                blkid_partition pp,
+                sd_device_enumerator *e,
+                Set **partnos,
+                RereadPartitionTableFlags flags,
+                bool *changed) {
+        int r;
+
+        assert(d);
+        assert(fd >= 0);
+        assert(pp);
+        assert(e);
+        assert(partnos);
+        assert(changed);
+
+        const char *node;
+        r = sd_device_get_devname(d, &node);
+        if (r < 0)
+                return log_device_debug_errno(d, r, "Failed to acquire device node path: %m");
+
+        errno = 0;
+        int nr = sym_blkid_partition_get_partno(pp);
+        if (nr < 0)
+                return log_debug_errno(errno_or_else(EIO), "Failed to read partition number of partition: %m");
+
+        log_device_debug(d, "Processing partition %i...", nr);
+
+        errno = 0;
+        blkid_loff_t start = sym_blkid_partition_get_start(pp);
+        if (start < 0)
+                return log_debug_errno(errno_or_else(EIO), "Failed to read partition start offset of partition %i: %m", nr);
+        assert((uint64_t) start < UINT64_MAX / 512U);
+
+        errno = 0;
+        blkid_loff_t size = sym_blkid_partition_get_size(pp);
+        if (size < 0)
+                return log_debug_errno(errno_or_else(EIO), "Failed to read partition size of partition %i: %m", nr);
+        assert((uint64_t) size < UINT64_MAX / 512U);
+
+        for (sd_device *partition = sd_device_enumerator_get_device_first(e);
+             partition;
+             partition = sd_device_enumerator_get_device_next(e)) {
+
+                unsigned nr_kernel;
+                r = device_get_property_uint(partition, "PARTN", &nr_kernel);
```

Comment:
Why not to generate device node path to the partition and acquire sd_device object from that? The loop for the enumerator looks not necessary.
```
partition_node_of(node, nr, &path);
sd_device_new_from_devname(&partition, path);
```